rllib/algorithms/sac/sac.py

```diff
@@ -371,3 +430,133 @@ def get_default_policy_class(
             return SACTorchPolicy
         else:
             return SACTFPolicy
+
+    @override(DQN)
+    def training_step(self) -> ResultDict:
+        # Check, if we use `RolloutWorker` or `EnvRunner`.
+        use_rollout_worker = self.config.env_runner_cls is None or issubclass(
+            self.config.env_runner_cls, RolloutWorker
+        )
+
+        # If `RolloutWorker` is used and the old stack, then use as before
+        # the training step of the DQN algorithm.
+        if use_rollout_worker or not self.config._enable_new_api_stack:
+            return super().training_step()
+
+        # Alternate between storing and sampling and training.
+        store_weight, sample_and_train_weight = calculate_rr_weights(self.config)
+        train_results = {}
+
+        # Run multiple sampling iterations.
+        for _ in range(store_weight):
+            # Otherwise, we use the new stack.
+            with self._timers[SAMPLE_TIMER]:
+                if self.workers.num_remote_workers() <= 0:
+                    episodes: List[SingleAgentEpisode] = [
+                        self.workers.local_worker().sample()
+                    ]
+                else:
+                    episodes: List[SingleAgentEpisode] = self.workers.foreach_worker(
+                        lambda w: w.sample(),
+                        local_worker=False,
+                    )
+                # Perform SAC postprocessing (Prio weights) on a (flattened)
+                # list of Episodes.
+                # TODO (simon): Change this to a preprocessing in the learner.
+                # postprocessed_episodes = self.postprocess_episodes(episodes)
+
+            episodes = tree.flatten(episodes)
+            # TODO (sven): single- vs multi-agent.
+            self._counters[NUM_AGENT_STEPS_SAMPLED] += sum(len(e) for e in episodes)
+            self._counters[NUM_ENV_STEPS_SAMPLED] += sum(len(e) for e in episodes)
+
+            # Add the sampled experiences to the replay buffer.
+            # TODO (simon): Apply the same function like in PPO to aggregate episodes
+            # over workers.
+            self.local_replay_buffer.add(episodes)
+
+        # Update the target network each `target_network_update_freq` steps.
+        current_ts = self._counters[
+            NUM_AGENT_STEPS_SAMPLED
+            if self.config.count_steps_by == "agent_steps"
+            else NUM_ENV_STEPS_SAMPLED
+        ]
+
+        # If enough experiences have been sampled start training.
+        if current_ts > self.config.num_steps_sampled_before_learning_starts:
+            # Run multiple training iterations.
+            for _ in range(sample_and_train_weight):
+                # Sample training batch from replay_buffer.
+                train_dict = self.local_replay_buffer.sample(
```

Comment:
Let's implement the n-step logic in the existing episode buffer (add a n-step arg to its constructor, allowing for single int or 2-tuple (pick uniformly from the range)).

Then the buffer would create the n-step samples on-the-fly during its `sample()` method: Pick an episode, pick a ts in the episode, pick a n-step value from the given range, get the obs at ts+n (this would be the NEXT_OBS), sum up all the rewards between ts and ts+n -> done.

Reply:
I think, we need to sum up rewards starting at index 1 instead of zero, correct? The reason for this is that the episode buffer samples for each observation the reward that came with it (or during the transfer to it), i.e. $(o_t, r_t, a_{t-1}, i_t$ is a single experience sampled in the replay buffer. 

So for 1-step we would need: $r_{t+1} + \gamma * Q(o_{t+1}, a_{t+1})$ and as we sample $a_{t+1}$ we need only a two consecutive timesteps sampled from the buffer.

Reply:
I think index 0 for rewards is still correct.

o0 = reset obs
ao = action taken after reset
r0 = reward received after taking action a0 (in state o0)
...

So the to-be-returned experience tuple from the buffer for n-step=2 would be:
`(obs=o0, act=a0, rew=r0+r1, next_obs=o2)`

Does this make sense?

For n-step=1 (default), this would simplify to the "normal" experience tuple:
`(obs=o0, act=a0, reward=r0, next_obs=o1)`

Reply:
Its correct that we have for an observation its corresponding action chosen in that state. 

I am not 100% sure, but I think when looking into lines [260](https://github.com/ray-project/ray/blob/525f2d7b5d11bb4b6f872ce5a4316036694678f5/rllib/utils/replay_buffers/episode_replay_buffer.py#L260) and [265](https://github.com/ray-project/ray/blob/525f2d7b5d11bb4b6f872ce5a4316036694678f5/rllib/utils/replay_buffers/episode_replay_buffer.py#L265) of the `EpisodeReplayBuffer` the reward is the one that came with the observation, i.e. for the reset obs `o0` it would be a dummy reward of `r0=0.0` like:

`o0` = reset obs
`a0` = action taken after reset
`r0` = dummy reward `0.0`
`o1` = obs after choosing in `o0` action `a0`
`a1` = action taken in `o1`
`r1` = reward received AFTER taking action `a0` in `o0`
...
`ot` = obs after choosing in `o(t-1)` action `a(t-1)`
`at` = action taken in `ot`
`rt` = reward received AFTER taking action `a(t-1)` in `o(t-1)`

----

Following this logic the buffer would need to return

`(obs=o0, act=a0, rew=r1 + r2, next_obs=o2)`

and for the "normal" 1-step:

`(obs=o0, act=a0, rew=r1, next_obs=o1)`

Reply:
@sven1977 As this is a crucial stage in all off-policy algorithms, can you take another look on the rewards in my logic above, pls? I think we have to take the next rewards when calculating e.g. targets in Q learning with this sampling in `EpisodeReplayBuffer`.

Reply:
You are right: This particular EpisodeReplayBuffer does indeed that (for DreamerV3's very specific needs), it adds a dummy reward (r0) right after o0 and the first actual reward received is then r1.

But either way, we should probably implement a sub-class of this buffer with its own `sample` method that abides more to SAC's (or off-policy algos in general) needs, e.g. including the n-step mechanism. What do you think? I feel like the current `EpisodeReplayBuffer`'s `sample()` method is too much geared toward DreamerV3's train batch requirements.

Reply:
Regarding your comments here and the one below. I would sit down tomorrow on #42369 and rewrite this sampling. I will reuse the `SegmentTree`s and try to keep as much of the logic in our `PriotizedReplayBuffer`.

Reply:
Okay the prioritized episode buffer with n-step sampling is ready for review now in PR #42832 . Let's combine all parts together and apply SAC on the tuned examples.